libr/util/skiplist.c

```diff
@@ -0,0 +1,260 @@
+// (c) 2016 Jeffrey Crowell, Riccardo Schirone(ret2libc)
+// BSD 3 Clause License
+// radare2
+
+// Skiplists are a probabilistic datastructure than can be used as a k-v store
+// with average case O(lg n) lookup time, and worst case O(n).
+
+// https://en.wikipedia.org/wiki/Skip_list
+
+#include <r_skiplist.h>
+
+const int kSkipListDepth = 15; // max depth
+
+static RSkipListNode *r_skiplist_node_new (void *data, int level) {
+	RSkipListNode *res = R_NEW (RSkipListNode);
+	if (!res) return NULL;
+	res->forward = R_NEWS0 (RSkipListNode *, level + 1);
+	if (!res->forward) goto err_forward;
+	res->data = data;
+	return res;
+
+err_forward:
+	free (res);
+	return NULL;
+}
+
+static void r_skiplist_node_free (RSkipList *list, RSkipListNode *node) {
+	if (list->freefn && node->data) {
```

Comment:
no need to check for node->data imho because the ->freefn will chk that already twice